mangaki/mangaki/utils/recommendations.py

```diff
@@ -5,119 +5,143 @@
+
+    filename = 'ratings.pickle'
+    if algo_name == 'knn' or not os.path.isfile('ratings.pickle'):
+        ratings_pack = make_anonymous_data(queryset)
+        if algo_name != 'knn':
+            with open(filename, 'wb') as f:
+                pickle.dump(ratings_pack, f)
     else:
-        banned_works = set(rated_works.keys())
-
-    mangas = Work.objects.filter(category__slug='manga')
-
-    if editor == 'otototaifu':
-        mangas = mangas.filter(editor__title__in=['Ototo Manga', 'Taifu comics'])
-    elif editor != 'unspecified':
-        mangas = mangas.filter(editor__title__icontains=editor)
-    manga_ids = mangas.values_list('id', flat=True)
-
-    kept_works = None
-    if category == 'anime':
-        banned_works |= set(manga_ids)
-    elif category == 'manga':
-        kept_works = set(manga_ids)
-
-    chrono.save('[%dQ] retrieve her %d ratings' % (len(connection.queries), len(rated_works)))
-    final_works = Counter()
-    nb_ratings = {}
-    c = 0
-    neighbors = Counter()
-    for user_id, work_id, choice in Rating.objects.filter(work__in=rated_works.keys()).values_list('user_id', 'work_id', 'choice'):
-        c += 1
-        neighbors[user_id] += values[rated_works[work_id]] * rating_values[choice]
-
-    chrono.save('[%dQ] fill neighbors with %d ratings' % (len(connection.queries), c))
-
-    score_of_neighbor = {}
-    # print('Neighbors:')
-    # nbr = []
-    for user_id, score in neighbors.most_common(NB_NEIGHBORS):
-        # print(User.objects.get(id=user_id).username, score)
-        score_of_neighbor[user_id] = score
-        # nbr.append(user_id)
-    # print(nbr)
-
-    sum_ratings = Counter()
-    nb_ratings = Counter()
-    sum_scores = Counter()
-    i = 0
-    for work_id, user_id, choice in Rating.objects.filter(user__id__in=score_of_neighbor.keys()).exclude(choice__in=['willsee', 'wontsee']).values_list('work_id', 'user_id', 'choice'):
-        i += 1
-        if work_id in banned_works or (kept_works and work_id not in kept_works):
-            continue
-
-        sum_ratings[work_id] += rating_values[choice]
-        nb_ratings[work_id] += 1
-        sum_scores[work_id] += score_of_neighbor[user_id]
+        with open(filename, 'rb') as f:
+            ratings_pack = pickle.load(f)
+    X, y, nb_users, nb_works, encode_user, decode_user, encode_work, decode_work, interesting_works = ratings_pack
+
+    chrono.save('[%dQ] get all %d interesting ratings' % (len(connection.queries), queryset.count()))
+
+    algo = ALGOS[algo_name]()
+    algo.set_parameters(nb_users, nb_works)
 
-    chrono.save('[%dQ] compute and filter all ratings from %d sources' % (len(connection.queries), i))
+    backup_filename = '%s-%s.pickle' % (algo.get_shortname(), datetime.strftime(datetime.now(), '%Y%m%d'))
+    if os.path.isfile(backup_filename):
+        algo.load(backup_filename)
+    else:
+        algo.fit(X, y)
+        if algo_name in {'svd', 'als'}:
```

Comment:
n = 2, right?
So tuple is accessed in O(1) right?

I checked the speed, no differences (using %timeit).
I checked the bytecode, no real differences except a tuple is "native":
- `0 LOAD_CONST               4 (('svd', 'als'))`
- `0 LOAD_CONST               3 (frozenset({'svd', 'als'}))`

Then, I checked the size taken in memory.
A frozenset seems to take 224 bytes on my computer.
A two-element tuple seems to take 64 bytes on my computer.

We could do even better if we wanted to nitpick by using enum integer values.
Honestly, I'd prefer tuple over set, in fact, it seems like it makes no difference until the day we need to optimise the hell out of Mangaki memory consumption in a future where we have 40M of users and we burn hardware as fast as [Dropbox](blog.kevmod.com/2017/02/personal-thoughts-about-pystons-outcome/) and goes on creating our own JIT by refusing to acknowledge PyPy existence.

Right. You can ignore my comment. Random rants of a young student.
(but please: tuples.)